doc/Subprocess debugging.md

```diff
@@ -0,0 +1,318 @@
+1. User code spawns child process.
+1. User stops debugging.
+
+```mermaid
+sequenceDiagram
+# Install "GitHub + Mermaid" from the Chrome Web Store to render the diagram
+
+participant IDE
+participant Adapter
+participant Debuggee_1
+participant Debuggee_2
+
+
+Note left of IDE: user starts<br/>debugging
+
+IDE ->> Adapter: spawn and connect over stdio
+
+IDE ->>+ Adapter: request "launch"
+
+Adapter ->>+ Debuggee_1: spawn and pass server listener port (cmdline)
+
+Debuggee_1 -->>- Adapter: connect to server listener port
+
+Adapter ->>+ Debuggee_1: request "initialize", "launch"
+
+Debuggee_1 -->>- Adapter: respond to "initialize", "launch"
+
+Debuggee_1 ->>+ Debuggee_1: start debug session
+
+Adapter -->>- IDE: respond to "launch"
+
+loop every message between IDE and Debuggee_1
+Note over IDE,Debuggee_1: propagate message
+end
+
+
+Note right of Debuggee_1: user code spawns<br/>child process
+
+Debuggee_1 ->>+ Debuggee_2: spawn and pass server listener port (cmdline)
+
+Debuggee_2 ->>- Adapter: connect to server listener port
+
+Adapter ->>+ Debuggee_2: request "pydevd_systemInfo"
+
+Debuggee_2 -->>- Adapter: respond to "pydevd_systemInfo"
+
+Adapter ->>+ IDE: "ptvsd_subprocess" event:
+
+IDE ->>- Adapter: connect to IDE listener port
+
+IDE ->>+ Adapter: request "attach" to Debuggee_2
+
+Adapter ->>+ Debuggee_2: request "initialize", "attach"
+
+Debuggee_2 -->>- Adapter: respond to "initialize", "attach"
+
+Debuggee_2 ->>+ Debuggee_2: start debug session
+
+Adapter -->>- IDE: respond to "attach"
+
+loop every message between IDE and Debuggee_2
+Note over IDE,Debuggee_2: propagate message
+end
+
+
+Note left of IDE: user stops debugging
+
+IDE ->>+ Adapter: request "disconnect" from Debuggee_2
+
+Adapter ->> Debuggee_2: request "disconnect"
+
+Debuggee_2 ->> Adapter: confirm "disconnect"
+
+Note over Adapter,Debuggee_2: TCP connection is maintained
+
+Adapter -->>- IDE: confirm "disconnect" from Debuggee_2
+
+IDE -X+ Adapter: request "disconnect" from Debuggee_1
+
+Note over Adapter: implies "terminate"
+
+Adapter -X+ Debuggee_1: request "terminate"
+
+Debuggee_1 -X Debuggee_2: kill process
+deactivate Debuggee_2
+
+Debuggee_1 -->>- Adapter: confirm "terminate"
+
+Debuggee_1 -X- Debuggee_1: exits
+
+Adapter -->>- IDE: confirm "disconnect" from Debuggee_1
+
+Adapter -X Adapter: exits
+```
+
+
+
+## "attach" scenario
+
+1. User starts debuggee process with debug server in it (ptvsd command line or `ptvsd.enable_attach()`).
+1. User starts debugging (F5) with "attach" debug config.
+1. User code spawns child process.
+1. User disconnects from debuggee.
+1. User reconnects to debuggee.
+
+```mermaid
+sequenceDiagram
+# Install "GitHub + Mermaid" from the Chrome Web Store to render the diagram
+
+participant IDE
+participant Adapter
+participant Debuggee_1
+participant Debuggee_2
+
+
+Note left of Debuggee_1: user spawns<br/>debuggee
+
+Debuggee_1 ->>+ Adapter: spawn and pass adapter listener port (cmdline)
```

Comment:
Should this diagram already handle https://github.com/microsoft/ptvsd/issues/1830?

Reply:
It originally captured the current state of affairs, so that we can see what exactly is missing - #1830 came out of that. I need to update the diagram for all the issues that we open.